src/ros_filter.cpp

```diff
@@ -1893,6 +1894,13 @@ void RosFilter<T>::periodicUpdate()
 
   rclcpp::Time cur_time = this->now();
 
+  const rclcpp::Duration update_timespan_lower_bound = update_timespan_ * 0.999;
+  if (update_timespan_lower_bound > cur_time - last_update_time_) {
```

Comment:
This is being called from a wall timer, why add this? 

If you're concerned about using a timer based on wall time vs simulation time, then we should be making a `timer_` based on the current clock. Checking this is an unacceptable hack if its goal is to deal with simulation time not running at 1.0x speed.

Reply:
> If you're concerned about using a timer based on wall time vs simulation time

That's correct. Specifically, using wall time may result in multiple `odometry/filtered` messages being published at the same simulation time, which may (and in some cases will) cause trouble downstream. Some SLAM packages are not particularly happy with is.

> then we should be making a timer_ based on the current clock. Checking this is an unacceptable hack if its goal is to deal with simulation time not running at 1.0x speed.

I absolutely agree with you, but unfortunately that's not possible in Dashing nor Eloquent, and it may or may not be possible in Foxy either -- a fix is currently in the [roadmap](https://index.ros.org/doc/ros2/Roadmap/#new-features) (see second bullet). `rclcpp` API is a bit deceiving in this regard, because while you can instantiate a timer that uses the current clock and [computes durations using that source](https://github.com/ros2/rcl/blob/41f4fba9664a3ed4a0ddc9985d510f04e10418a4/rcl/src/rcl/wait.c#L568-L570), under the hood it sleeps as if [that duration was specified in system time](https://github.com/ros2/rcl/blob/41f4fba9664a3ed4a0ddc9985d510f04e10418a4/rcl/src/rcl/wait.c#L606-L613). So the `RTF >= 1.0` precondition remains. FWIW, [ROS1 solution wasn't much better](https://github.com/ros/ros_comm/blob/c16649772847c360d6b5275f2d54a2a728f49d8d/clients/roscpp/include/ros/timer_manager.h#L587-L590).

Having said that, I can change the code to use a proper timer so that when the feature lands we can drop these additional checks. WDYT?

Reply:
I think revert and wait for the feature to be properly supported. Its annoying  but if we merge this, that hack will never go away because functionally we'll all forget and then there will be a weird cause of non-determinism in simulation

Reply:
Done in https://github.com/cra-ros-pkg/robot_localization/pull/536/commits/d5629d4a1571205df7412d219e4b953862c78818.

Reply:
Oh, I just see your last comment. The last commit is as correct as it can be, and it'll remain correct once the feature is properly supported. You do have to bear in mind not to use RTF < 1.0, but that's true for every node.